--- a/CMakeLists.txt	2020-03-26 12:43:53.296075681 +0100
+++ b/CMakeLists.txt	2020-03-26 12:44:03.376075962 +0100
@@ -82,7 +82,7 @@ set_package_properties("libssh" PROPERTI
 )
 
 # Needs porting
-find_program(FREERDP_EXECUTABLE xfreerdp)
+find_program(FREERDP_EXECUTABLE xfreerdp2)
 if(FREERDP_EXECUTABLE)
     set(FREERDP_FOUND true) # mark the package found
     set(FREERDP_EXECUTABLE_FOUND true)
--- a/rdp/rdpview.cpp	2020-03-26 12:44:41.758077033 +0100
+++ b/rdp/rdpview.cpp	2020-03-26 12:45:53.268079028 +0100
@@ -160,7 +160,7 @@ bool RdpView::start()
 
     // Check the version of FreeRDP so we can use pre-1.1 switches if needed
     QProcess *xfreeRDPVersionCheck = new QProcess(this);
-    xfreeRDPVersionCheck->start(QStringLiteral("xfreerdp"), QStringList(QStringLiteral("--version")));
+    xfreeRDPVersionCheck->start(QStringLiteral("xfreerdp2"), QStringList(QStringLiteral("--version")));
     xfreeRDPVersionCheck->waitForFinished();
     QString versionOutput = QString::fromUtf8(xfreeRDPVersionCheck->readAllStandardOutput().constData());
     xfreeRDPVersionCheck->deleteLater();
@@ -373,10 +373,10 @@ bool RdpView::start()
     connect(m_process, SIGNAL(started()), SLOT(connectionOpened()));
 
     if (KWindowSystem::platform() == KWindowSystem::Platform::Wayland) {
-        m_process->start(QStringLiteral("wlfreerdp"), arguments);
+        m_process->start(QStringLiteral("wlfreerdp2"), arguments);
     }
     else {
-        m_process->start(QStringLiteral("xfreerdp"), arguments);
+        m_process->start(QStringLiteral("xfreerdp2"), arguments);
     }
 
     return true;
@@ -447,11 +447,11 @@ void RdpView::processError(QProcess::Pro
             QString executable;
             switch (KWindowSystem::platform()) {
             case KWindowSystem::Platform::Wayland:
-                executable = QStringLiteral("wlfreerdp");
+                executable = QStringLiteral("wlfreerdp2");
                 break;
             case KWindowSystem::Platform::X11:
             case KWindowSystem::Platform::Unknown:
-                executable = QStringLiteral("xfreerdp");
+                executable = QStringLiteral("xfreerdp2");
             }
             connectionError(i18n("Could not start \"%1\"; make sure %1 is properly installed.", executable),
                             i18n("RDP Failure"));
