--- a/meson.build	2022-01-09 00:22:38.722991810 +0100
+++ b/meson.build	2022-01-09 00:23:20.042991416 +0100
@@ -92,7 +92,7 @@ gmodule_dep = dependency('gmodule-2.0',
 libfprint_dep = dependency('libfprint-2', version: '>=' + libfprint_min_version)
 polkit_gobject_dep = dependency('polkit-gobject-1', version: '>= 0.91')
 dbus_dep = dependency('dbus-1', required: false)
-libsystemd_dep = dependency('libsystemd', required: get_option('pam'))
+libsystemd_dep = dependency('libsystemd', required: get_option('systemd'))
 pam_dep = cc.find_library('pam',
     required: get_option('pam'),
     has_headers: 'security/pam_modules.h',
--- a/pam/meson.build	2021-11-03 15:34:26.000000000 +0100
+++ b/pam/meson.build	2022-01-10 23:59:39.302357016 +0100
@@ -4,6 +4,8 @@ if pam_modules_dir == ''
     pam_modules_dir = '/' / get_option('libdir') / 'security'
 endif
 
+dbusg_dep = dependency('dbus-glib-1', required: true)
+
 pam_fprintd = shared_module('pam_fprintd',
     name_prefix: '',
     include_directories: [
@@ -14,7 +17,8 @@ pam_fprintd = shared_module('pam_fprintd
         'fingerprint-strings.h',
     ],
     dependencies: [
-        libsystemd_dep,
+        glib_dep,
+	dbusg_dep,
         pam_dep,
     ],
     c_args: [
