--- a/openbsd-compat/Makefile.am	2019-11-08 14:20:51.967929549 +0100
+++ b/openbsd-compat/Makefile.am	2019-11-08 14:21:06.961262888 +0100
@@ -1,7 +1,7 @@
 noinst_LIBRARIES = libopenbsd-compat.a
 
 libopenbsd_compat_a_SOURCES =	arc4random.c
-libopenbsd_compat_a_SOURCES +=	SSL_CTX_use_certificate_chain_mem.c
+#libopenbsd_compat_a_SOURCES +=	SSL_CTX_use_certificate_chain_mem.c
 libopenbsd_compat_a_SOURCES +=	base64.c
 libopenbsd_compat_a_SOURCES +=	bsd-getpeereid.c
 libopenbsd_compat_a_SOURCES +=	bsd-misc.c
