diff --git a/folly/experimental/logging/example/main.cpp b/folly/experimental/logging/example/main.cpp
index 14d08836aa..be4ba77a74 100644
--- a/third-party/folly/src/folly/experimental/logging/example/main.cpp
+++ b/third-party/folly/src/folly/experimental/logging/example/main.cpp
@@ -16,6 +16,7 @@
 #include <folly/experimental/logging/Init.h>
 #include <folly/experimental/logging/xlog.h>
 #include <folly/init/Init.h>
+#include <folly/portability/GFlags.h>
 
 #include <folly/experimental/logging/example/lib.h>
 
