--- a/third-party/re2/src/re2/dfa.cc.orig	2017-10-16 17:43:26.000000000 +0200
+++ b/third-party/re2/src/re2/dfa.cc	2017-10-23 12:45:26.823333335 +0200
@@ -101,7 +101,7 @@ class DFA {
     uint flag_;         // Empty string bitfield flags in effect on the way
                         // into this state, along with kFlagMatch if this
                         // is a matching state.
-    std::atomic<State*> next_[];    // Outgoing arrows from State,
+    std::atomic<State*> next_[0];    // Outgoing arrows from State,
                         // one per input byte class
   };
 
