--- a/src/MEGASync/mega/src/gfx/freeimage.cpp.old	2022-10-30 21:31:53.816333597 +0100
+++ b/src/MEGASync/mega/src/gfx/freeimage.cpp	2022-10-30 21:39:59.862347155 +0100
@@ -313,7 +313,7 @@ bool GfxProviderFreeImage::readbitmapFfm
 
     // Find decoder for video stream
     AVCodecID codecId = codecParm->codec_id;
-    AVCodec* decoder = avcodec_find_decoder(codecId);
+    const AVCodec* decoder = avcodec_find_decoder(codecId);
     if (!decoder)
     {
         LOG_warn << "Codec not found: " << codecId;
@@ -329,8 +329,7 @@ bool GfxProviderFreeImage::readbitmapFfm
     }
 
     // Force seeking to key frames
-    formatContext->seek2any = false;
-    videoStream->skip_to_keyframe = true;
+    formatContext->seek2any = true;
     if (decoder->capabilities & CAP_TRUNCATED)
     {
         codecContext->flags |= CAP_TRUNCATED;
